mcp/src/test/java/io/modelcontextprotocol/client/LifecyleInitializerTests.java

```diff
@@ -0,0 +1,403 @@
+			.verify();
+	}
+
+	@Test
+	void shouldReuseExistingInitialization() {
+		// First initialization
+		StepVerifier.create(initializer.withIntitialization("test1", init -> Mono.just("result1")))
+			.expectNext("result1")
+			.verifyComplete();
+
+		// Second call should reuse the same initialization
+		StepVerifier.create(initializer.withIntitialization("test2", init -> Mono.just("result2")))
+			.expectNext("result2")
+			.verifyComplete();
+
+		// Verify session was created only once
+		verify(mockSessionSupplier, times(1)).apply(any(ContextView.class));
+		verify(mockSession, times(1)).sendRequest(eq(McpSchema.METHOD_INITIALIZE), any(), any());
+	}
+
+	@Test
+	void shouldHandleConcurrentInitializationRequests() {
+		AtomicInteger sessionCreationCount = new AtomicInteger(0);
+
+		when(mockSessionSupplier.apply(any(ContextView.class))).thenAnswer(invocation -> {
+			sessionCreationCount.incrementAndGet();
+			return mockSession;
+		});
+
+		// Start multiple concurrent initializations
+		Mono<String> init1 = initializer.withIntitialization("test1", init -> Mono.just("result1"));
+		Mono<String> init2 = initializer.withIntitialization("test2", init -> Mono.just("result2"));
+		Mono<String> init3 = initializer.withIntitialization("test3", init -> Mono.just("result3"));
+
+		StepVerifier.create(Mono.zip(init1, init2, init3)).assertNext(tuple -> {
+			assertThat(tuple.getT1()).isEqualTo("result1");
+			assertThat(tuple.getT2()).isEqualTo("result2");
+			assertThat(tuple.getT3()).isEqualTo("result3");
+		}).verifyComplete();
+
+		// Should only create one session despite concurrent requests
+		assertThat(sessionCreationCount.get()).isEqualTo(1);
+		verify(mockSession, times(1)).sendRequest(eq(McpSchema.METHOD_INITIALIZE), any(), any());
+	}
+
+	@Test
+	void shouldHandleInitializationFailure() {
+		when(mockSession.sendRequest(eq(McpSchema.METHOD_INITIALIZE), any(), any()))
+			.thenReturn(Mono.error(new RuntimeException("Connection failed")));
+
+		StepVerifier.create(initializer.withIntitialization("test", init -> Mono.just(init.initializeResult())))
+			.expectError(McpError.class)
+			.verify();
+
+		assertThat(initializer.isInitialized()).isFalse();
+		assertThat(initializer.currentInitializationResult()).isNull();
+	}
+
+	@Test
+	void shouldHandleTransportSessionNotFoundException() {
+		// Simulate a successful initialization first
+		StepVerifier.create(initializer.withIntitialization("test", init -> Mono.just(init.initializeResult())))
+			.expectNext(MOCK_INIT_RESULT)
+			.verifyComplete();
+
+		assertThat(initializer.isInitialized()).isTrue();
+
+		// Simulate transport session not found exception
+		initializer.handleException(new McpTransportSessionNotFoundException("Session not found"));
+
+		// The exception handling resets the initialization state and triggers
+		// re-initialization
+		// We need to wait a bit for the async re-initialization to start
+		try {
+			Thread.sleep(10); // Small delay to allow async processing
```

Comment:
Every time we use sleep we almost certainly introduce flakiness into the test suite, please let's avoid it.